cql/src/main/java/org/apache/cassandra/stargate/transport/internal/PlainTextTokenSaslNegotiator.java

```diff
@@ -0,0 +1,127 @@
+package org.apache.cassandra.stargate.transport.internal;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.stargate.auth.AuthenticationService;
+import io.stargate.auth.StoredCredentials;
+import io.stargate.db.AuthenticatedUser;
+import io.stargate.db.Authenticator;
+import io.stargate.db.Persistence;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import org.apache.cassandra.stargate.exceptions.AuthenticationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class PlainTextTokenSaslNegotiator implements Authenticator.SaslNegotiator {
+  private static final Logger logger = LoggerFactory.getLogger(PlainTextTokenSaslNegotiator.class);
+
+  static final String TOKEN_USERNAME = System.getProperty("stargate.cql_token_username", "token");
+  static final int TOKEN_MAX_LENGTH =
+      Integer.parseInt(System.getProperty("stargate.cql_token_max_length", "36"));
+
+  static final byte NUL = 0;
+
+  private final Persistence<?, ?, ?> persistence;
+  private final AuthenticationService authentication;
+  private final Authenticator.SaslNegotiator wrapped;
+  private StoredCredentials storedCredentials;
+
+  PlainTextTokenSaslNegotiator(
+      Authenticator.SaslNegotiator wrapped,
+      Persistence<?, ?, ?> persistence,
+      AuthenticationService authentication) {
+    this.persistence = persistence;
+    this.authentication = authentication;
+    this.wrapped = wrapped;
+  }
+
+  @Override
+  public byte[] evaluateResponse(byte[] clientResponse) throws AuthenticationException {
+    if (attemptTokenAuthentication(clientResponse)) return null;
+    return wrapped.evaluateResponse(clientResponse);
+  }
+
+  @Override
+  public boolean isComplete() {
+    return storedCredentials != null || wrapped.isComplete();
+  }
+
+  @Override
+  public AuthenticatedUser<?> getAuthenticatedUser() throws AuthenticationException {
+    if (storedCredentials != null)
+      return persistence.newAuthenticatedUser(storedCredentials.getRoleName());
+    else return wrapped.getAuthenticatedUser();
+  }
+
+  @VisibleForTesting
+  boolean attemptTokenAuthentication(byte[] clientResponse) {
+    try {
+      Credentials credentials = decodeCredentials(clientResponse);
+
+      if (!credentials.username.equals(TOKEN_USERNAME)) return false;
```

Comment:
Can we guarantee username is not null?

Reply:
Yes: decodeCredentials throws if the user is null.